src/style/style.js

```diff
@@ -1497,6 +1497,10 @@ class Style extends Evented {
         } else { // Updating
             const terrain = this.terrain;
             const currSpec = terrain.get();
+
+            // Add in terrain default values if it was not set in style
+            if (!terrainOptions.hasOwnProperty('exaggeration')) terrainOptions.exaggeration = 1;
```

Comment:
I think we probably shouldn't have a hardcoded exception here like this. Is this still necessary now that you've switched from a shallow clone to a deep clone?

Reply:
Two things to note for that:
1. It might be more robust to access the default value from the spec, something like:
```
import styleSpec from '../style-spec/reference/latest.js';
...
styleSpec.terrain['exaggeration'].default
```
2. We have a slightly similar code path for fog: https://github.com/mapbox/mapbox-gl-js/blob/main/src/style/fog.js#L82_L87, so maybe this code could be moved [here](https://github.com/mapbox/mapbox-gl-js/blob/main/src/style/terrain.js#L47) when setting terrain, and loop through all style spec properties, so that adding new terrain specification properties doesn't break this code and these two root properties would have similar logic

Reply:
Ah thanks @karimnaaji  - I tried finding a way to not have to hardcode this value as a last resort. Much appreciated!

Reply:
It made more sense to put the loop [here](https://github.com/mapbox/mapbox-gl-js/pull/11942/files#diff-c19065149b8e6b02ea30bd2576429dfc2f917695da9382c77a6957401f6fef4eR1501). In `setTerrain` it's looping through the new terrain options and only updating the new properties that aren't equal to the old terrain's options, so if the default property+value isn't added to the new terrain options before this, it won't be passed through to `terrain.set`.

Reply:
@avpeery Do you think that we should also move it [there](https://github.com/mapbox/mapbox-gl-js/blob/main/src/style/style.js#L1567) for fog? If so, maybe there's a way we could consolidate root properties updates and make share this code between the two.